File: public/api/generate.php
<?php

require dirname(__DIR__, 2) . '/vendor/autoload.php';

use App\Database;
use App\ImageRepository;
use App\CropRepository;
use App\PresetLoader;
use App\GlideServer;
use App\AppConfig;
use App\ComponentLoader;

header('Content-Type: application/json');

if ($_SERVER['REQUEST_METHOD'] !== 'POST') {
    http_response_code(405);
    echo json_encode(['error' => 'Method not allowed']);
    exit;
}

$body        = json_decode(file_get_contents('php://input'), true);
$imageId     = $body['image_id']    ?? '';
$presetId    = $body['preset_id']   ?? '';
$variantId   = $body['variant_id']  ?? null;
$componentId = $body['component_id'] ?? null;

if ($imageId === '' || $presetId === '') {
    http_response_code(400);
    echo json_encode(['error' => 'Missing image_id or preset_id']);
    exit;
}

$db     = Database::get();
$images = new ImageRepository($db);
$crops  = new CropRepository($db);

$image = $images->findById($imageId);
if ($image === null) {
    http_response_code(404);
    echo json_encode(['error' => 'Image not found']);
    exit;
}

$preset = PresetLoader::findPreset($presetId);
if ($preset === null) {
    http_response_code(400);
    echo json_encode(['error' => 'Unknown preset']);
    exit;
}

$outputFolder  = AppConfig::outputFolder();
$defaultFormat = AppConfig::defaultFormat();

if ($outputFolder === '' || !is_dir($outputFolder)) {
    http_response_code(400);
    echo json_encode(['error' => 'Output folder not configured or does not exist']);
    exit;
}

// Build output directory: {outputFolder}/{component_id}/{image_slug}/
// Falls back to {outputFolder}/{image_slug}/ when no component context
$imageSlug = slugify($image['name'] ?: $image['original_name']);
$subDir    = $componentId !== null
    ? $outputFolder . '/' . $componentId . '/' . $imageSlug
    : $outputFolder . '/' . $imageSlug;

if (!is_dir($subDir)) {
    mkdir($subDir, 0755, true);
}

$server    = GlideServer::create();
$generated = [];

$variants = $variantId !== null
    ? array_filter($preset['variants'], fn($v) => $v['id'] === $variantId)
    : $preset['variants'];

foreach ($variants as $variant) {
    $savedCrop = $crops->find($imageId, $presetId, $variant['id']);

    if ($savedCrop !== null) {
        $cropX = (int) $savedCrop['crop_x'];
        $cropY = (int) $savedCrop['crop_y'];
        $cropW = (int) $savedCrop['crop_w'];
        $cropH = (int) $savedCrop['crop_h'];
    } else {
        $ratio = $variant['width'] / $variant['height'];
        $srcW  = $image['width'];
        $srcH  = $image['height'];

        if ($srcW / $srcH > $ratio) {
            $cropH = $srcH;
            $cropW = (int) round($srcH * $ratio);
        } else {
            $cropW = $srcW;
            $cropH = (int) round($srcW / $ratio);
        }

        $cropX = (int) round(($srcW - $cropW) / 2);
        $cropY = (int) round(($srcH - $cropH) / 2);
    }

    $format  = $variant['format'] ?? $preset['format'] ?? $defaultFormat;
    $cropStr = "{$cropW},{$cropH},{$cropX},{$cropY}";
    $base    = $presetId . '_' . $variant['id'];
    $files   = [];

    foreach ([1, 2] as $dpr) {
        $w       = $variant['width']  * $dpr;
        $h       = $variant['height'] * $dpr;
        $suffix  = $dpr === 1 ? '1x' : '2x';
        $outName = $base . '_' . $suffix . '.' . $format;
        $outPath = $subDir . '/' . $outName;

        $params = [
            'w'    => $w,
            'h'    => $h,
            'fit'  => 'crop',
            'crop' => $cropStr,
            'fm'   => $format,
        ];

        $cachePath = $server->makeImage($image['filename'], $params);
        $contents  = $server->getCache()->read($cachePath);
        file_put_contents($outPath, $contents);

        $files[$suffix] = $outName;
    }

    $crops->upsert($imageId, $presetId, $variant['id'], $cropX, $cropY, $cropW, $cropH);
    $crops->markGenerated($imageId, $presetId, $variant['id']);

    $generated[] = [
        'variant_id' => $variant['id'],
        'files'      => $files,
        'folder'     => $subDir,
    ];
}

echo json_encode(['generated' => $generated]);

File: internal/handler/generate.go
package handler

import (
	"fmt"
	imgpkg "imagemanager/internal/imaging"
	"imagemanager/internal/config"
	"imagemanager/internal/model"
	"imagemanager/internal/repo"
	"math"
	"net/http"
	"os"
	"path/filepath"
)

func (a *App) Generate(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		jsonErr(w, 405, "Method not allowed")
		return
	}

	var body struct {
		ImageID     string  `json:"image_id"`
		PresetID    string  `json:"preset_id"`
		VariantID   *string `json:"variant_id"`
		ComponentID *string `json:"component_id"`
	}
	if err := readJSON(r, &body); err != nil {
		jsonErr(w, 400, "Invalid request body")
		return
	}

	if body.ImageID == "" || body.PresetID == "" {
		jsonErr(w, 400, "Missing image_id or preset_id")
		return
	}

	imgRepo := repo.NewImageRepo(a.DB)
	cropRepo := repo.NewCropRepo(a.DB)

	image, err := imgRepo.FindByID(body.ImageID)
	if err != nil || image == nil {
		jsonErr(w, 404, "Image not found")
		return
	}

	presets, err := config.LoadPresets(a.AppRoot)
	if err != nil {
		jsonErr(w, 500, "Failed to load presets")
		return
	}
	preset := config.FindPreset(presets, body.PresetID)
	if preset == nil {
		jsonErr(w, 400, "Unknown preset")
		return
	}

	cfg, err := config.LoadAppConfig(a.AppRoot)
	if err != nil {
		jsonErr(w, 500, "Failed to load config")
		return
	}

	if cfg.OutputFolder == "" {
		jsonErr(w, 400, "Output folder not configured")
		return
	}
	if _, err := os.Stat(cfg.OutputFolder); os.IsNotExist(err) {
		jsonErr(w, 400, "Output folder does not exist")
		return
	}

	imageSlug := config.Slugify(image.Name)
	componentID := ""
	if body.ComponentID != nil {
		componentID = *body.ComponentID
	}

	var subDir string
	if componentID != "" {
		subDir = filepath.Join(cfg.OutputFolder, componentID, imageSlug)
	} else {
		subDir = filepath.Join(cfg.OutputFolder, imageSlug)
	}

	if err := os.MkdirAll(subDir, 0755); err != nil {
		jsonErr(w, 500, "Failed to create output directory")
		return
	}

	// Load source image once
	src, err := loadSourceImage(a.AppRoot, image.Filename, image.MimeType)
	if err != nil {
		jsonErr(w, 500, "Failed to load source image")
		return
	}

	// Select variants to process
	variants := preset.Variants
	if body.VariantID != nil && *body.VariantID != "" {
		var onlyVariants []model.Variant
		for _, v := range preset.Variants {
			if v.ID == *body.VariantID {
				onlyVariants = append(onlyVariants, v)
			}
		}
		variants = onlyVariants
	}

	type generatedEntry struct {
		VariantID string            `json:"variant_id"`
		Files     map[string]string `json:"files"`
		Folder    string            `json:"folder"`
	}

	var generated []generatedEntry

	for _, variant := range variants {
		// Get or auto-center crop
		savedCrop, _ := cropRepo.Find(body.ImageID, body.PresetID, variant.ID)

		var cropX, cropY, cropW, cropH int
		if savedCrop != nil {
			cropX = savedCrop.CropX
			cropY = savedCrop.CropY
			cropW = savedCrop.CropW
			cropH = savedCrop.CropH
		} else {
			ratio := float64(variant.Width) / float64(variant.Height)
			srcW := float64(image.Width)
			srcH := float64(image.Height)

			if srcW/srcH > ratio {
				cropH = image.Height
				cropW = int(math.Round(srcH * ratio))
			} else {
				cropW = image.Width
				cropH = int(math.Round(srcW / ratio))
			}
			cropX = int(math.Round((srcW - float64(cropW)) / 2))
			cropY = int(math.Round((srcH - float64(cropH)) / 2))
		}

		format := variant.Format
		if format == "" {
			format = preset.Format
		}
		if format == "" {
			format = cfg.DefaultFormat
		}

		ext := imgpkg.OutputExt(format)
		base := body.PresetID + "_" + variant.ID
		files := make(map[string]string)

		for _, dpr := range []int{1, 2} {
			targetW := variant.Width * dpr
			targetH := variant.Height * dpr
			suffix := fmt.Sprintf("%dx", dpr)
			outName := base + "_" + suffix + "." + ext
			outPath := filepath.Join(subDir, outName)

			out := imgpkg.CropAndResize(src, cropX, cropY, cropW, cropH, targetW, targetH)

			outFile, err := os.Create(outPath)
			if err != nil {
				jsonErr(w, 500, "Failed to create output file")
				return
			}
			if err := imgpkg.EncodeImage(outFile, out, format); err != nil {
				outFile.Close()
				jsonErr(w, 500, "Failed to encode output image")
				return
			}
			outFile.Close()

			files[suffix] = outName
		}

		cropRepo.Upsert(body.ImageID, body.PresetID, variant.ID, cropX, cropY, cropW, cropH)
		cropRepo.MarkGenerated(body.ImageID, body.PresetID, variant.ID)

		generated = append(generated, generatedEntry{
			VariantID: variant.ID,
			Files:     files,
			Folder:    subDir,
		})
	}

	jsonOK(w, map[string]any{"generated": generated})
}

